46-sugar-ant/technical_specification.md

```diff
@@ -0,0 +1,300 @@
+
+The internal auth token will be only created and passed on by the Auth Adapter if the use is fully authenticated, i.e. logged in via LS Login and the second factor has been validated.
+
+The `status` field will be removed from the auth token. The existence of the token always implies that the user account is active and not invalid.
+
+The `ext_id` field will be removed from the auth token. There are only two exceptional cases where it is needed, and in theses cases it can be stored in the `id` field instead. The `id` field will also be made mandatory and required to be a non-empty string.
+
+There are only the following three exceptions where the auth token will be also added by the Auth Adapter if the user is only logged in via LS Login:
+
+- `POST /users`
+	- used to self-register a user
+	- the `id` field of the auth context will contain the external id, not the internal id
+- `PUT /users/{user_id}`
+	- when requested by users to confirm a name an email change
+	- the user must be already registered
+	- the `id` field of the auth context will contain the internal id of the user, and it must correspond to the `user_id` in the path
+- `GET /users/{ext_id}`
+	- when requested by users to check their own registered data
+	- the `id` field of the auth context will contain the external id, not the internal id, and it must correspond to the `ext_id` in the path
+
+### Service Commons Library
+
+The `ghga-service-commons` library must be changed to reflect the changes in the internal auth token as outlined above.
+
+### Backend Models
+
+A new `IVA` (independent verification address) model must be added to the User Management service:
+	- `id`: string (unique internal id)
+	- `user_id`: string (internal id of the user)
+	- `type`: enum (phone, fax, postal address, in-person)
+	- `value`: string (the actual phone number)
+	- `verification_code`:  string (random number)
+	- `state`: enum (unverified, requested, pending, verified)
+
+The `IVA`s should be maintained in a separate collection by the User Management service.
+
+The `Claims` model must be extended so that claims in addition to referencing a user, it can optionally also reference an `IVA` via an additional property `iva_id`.
+
+A new `TOTPToken` model must be created that includes all attributes required to operate a TOTP token, like the OTP secret key that it is based upon (details will be defined during implementation).
+
+The `TOTPToken`s will be managed by the Auth Adapter, as explained above.
+
+The `AccessRequest` model must be extended to also include an optional `iva_id` field referencing an IVA that is used to verify the corresponding access grant. This field will be initially set to `None`, and it will be set after access has been allowed.
+
+### Login Flow in the Frontend
+
+The frontend keeps the state of the current user in the session storage. The in the frontend can be one of the following stages:
+- `unauthenticated`
+- `logged-in` (with one factor)
```

Comment:
Maybe name differently

Reply:
Ok, renamed to not cause confusion with different naming in whitepaper.